Do warmup in cudagraph pool

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #96775
* __->__ #96433
* #96384
* #96154
* #89146
* #95020
* #94943
* #94653


The main change here is to do warmup within a cudagraph pool. If you imagine we are part of the way through training, all of the memory would be allocated to the CUDAGraph pool, and if we were to try to run a model again in the eager pool we would OOM. Moreover, for many models you need to reclaim the activations in the backward as you run it, otherwise you have a steep memory penalty, so keeping the inputs alive as you do warmup is not an option. 

The new flow is:
- Warmup and return outputs
- Record and run graph
- Replay Graph. 

There are a few smaller changes that are in this PR.
- Have a separate TapeContainerManager per device. This follows the same per device logic of the Caching Allocator. Also limits CUDAGraphs in the multi device scenario. TBH, i'm not sure how this would work even for the other more naive implementation. 
- Simplifies pool preservation by having the TapeContainer store a graph. 
- Also fixed one bug where we were keeping an output that was an alias of a parameter as a path_live_weakref, which would then get freed when we needed to record a new generation and it was still live.





cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

Differential Revision: [D43999885](https://our.internmc.facebook.com/intern/diff/D43999885)

eellison committed:
[WIP] Do warmup in cudagraph pool

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Do warmup in cudagraph pool"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "[WIP] Do warmup in cudagraph pool"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Do warmup in cudagraph pool"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Do warmup in cudagraph pool"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Do warmup in cudagraph pool"




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

eellison committed:
Update on "Do warmup in cudagraph pool"



This is working but probably needs a bit of clean up before review.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D43999885).

eellison committed:
Update on "Do warmup in cudagraph pool"



This is working but probably needs a bit of clean up before review.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D43999885](https://our.internmc.facebook.com/intern/diff/D43999885)

[ghstack-poisoned]

eellison committed:
Update on "Do warmup in cudagraph pool"



This is working but probably needs a bit of clean up before review.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D43999885](https://our.internmc.facebook.com/intern/diff/D43999885)

[ghstack-poisoned]

eellison committed:
Update on "Do warmup in cudagraph pool"



This is working but probably needs a bit of clean up before review.

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D43999885](https://our.internmc.facebook.com/intern/diff/D43999885)

[ghstack-poisoned]

eellison committed:
Update on "Do warmup in cudagraph pool"



The main change here is to do warmup within a cudagraph pool. If you imagine we are part of the way through training, all of the memory would be allocated to the CUDAGraph pool, and if we were to try to run a model again in the eager pool we would OOM. Moreover, for many models you need to reclaim the activations in the backward as you run it, otherwise you have a steep memory penalty, so keeping the inputs alive as you do warmup is not an option. 

The new flow is:
- Warmup and return outputs
- Record and run graph
- Replay Graph. 

There are a few smaller changes that are in this PR.
- Have a separate TapeContainerManager per device. This follows the same per device logic of the Caching Allocator. Also limits CUDAGraphs in the multi device scenario. TBH, i'm not sure how this would work even for the other more naive implementation. 
- Simplifies pool preservation by having the TapeContainer store a graph. 
- Also fixed one bug where we were keeping an output that was an alias of a parameter as a path_live_weakref, which would then get freed when we needed to record a new generation and it was still live.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D43999885](https://our.internmc.facebook.com/intern/diff/D43999885)

[ghstack-poisoned]

eellison committed:
Update on "Do warmup in cudagraph pool"



The main change here is to do warmup within a cudagraph pool. If you imagine we are part of the way through training, all of the memory would be allocated to the CUDAGraph pool, and if we were to try to run a model again in the eager pool we would OOM. Moreover, for many models you need to reclaim the activations in the backward as you run it, otherwise you have a steep memory penalty, so keeping the inputs alive as you do warmup is not an option. 

The new flow is:
- Warmup and return outputs
- Record and run graph
- Replay Graph. 

There are a few smaller changes that are in this PR.
- Have a separate TapeContainerManager per device. This follows the same per device logic of the Caching Allocator. Also limits CUDAGraphs in the multi device scenario. TBH, i'm not sure how this would work even for the other more naive implementation. 
- Simplifies pool preservation by having the TapeContainer store a graph. 
- Also fixed one bug where we were keeping an output that was an alias of a parameter as a path_live_weakref, which would then get freed when we needed to record a new generation and it was still live.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D43999885](https://our.internmc.facebook.com/intern/diff/D43999885)

[ghstack-poisoned]

eellison committed:
Update on "Do warmup in cudagraph pool"



The main change here is to do warmup within a cudagraph pool. If you imagine we are part of the way through training, all of the memory would be allocated to the CUDAGraph pool, and if we were to try to run a model again in the eager pool we would OOM. Moreover, for many models you need to reclaim the activations in the backward as you run it, otherwise you have a steep memory penalty, so keeping the inputs alive as you do warmup is not an option. 

The new flow is:
- Warmup and return outputs
- Record and run graph
- Replay Graph. 

There are a few smaller changes that are in this PR.
- Have a separate TapeContainerManager per device. This follows the same per device logic of the Caching Allocator. Also limits CUDAGraphs in the multi device scenario. TBH, i'm not sure how this would work even for the other more naive implementation. 
- Simplifies pool preservation by having the TapeContainer store a graph. 
- Also fixed one bug where we were keeping an output that was an alias of a parameter as a path_live_weakref, which would then get freed when we needed to record a new generation and it was still live.





cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

Differential Revision: [D43999885](https://our.internmc.facebook.com/intern/diff/D43999885)

[ghstack-poisoned]

eellison commented:
Squashed into https://github.com/pytorch/pytorch/pull/89146
